src/esp/assets/GenericSemanticMeshData.cpp

```diff
@@ -301,64 +269,68 @@ GenericSemanticMeshData::buildSemanticMeshData(
       objectIds = std::move(out.first);
       auto clrMapView = colorsThatBecomeTheColorMap.prefix(out.second);
       colorMapToUse.assign(clrMapView.begin(), clrMapView.end());
-
-      objPartitionsFromSSD = false;
     }
   }
-  if (!objPartitionsFromSSD) {
-    // if objPartitionsFromSSD then semanticData->objectIds were
-    // populated directly
+  if (!semanticMeshData->meshHasSeparatePartitionIDs) {
+    // if objPartitionsFromSSD then semanticMeshData->objectIds were
+    // populated directly while partition IDs were derived
     Mn::Math::castInto(
         Cr::Containers::arrayCast<2, Mn::UnsignedInt>(
             Cr::Containers::stridedArrayView(objectIds)),
         Cr::Containers::arrayCast<2, Mn::UnsignedShort>(
-            Cr::Containers::stridedArrayView(semanticData->objectIds_)));
+            Cr::Containers::stridedArrayView(semanticMeshData->objectIds_)));
+  }
+
+  semanticMeshData->meshHasPartitionIDXs =
+      (objIdsFromSrcMesh || semanticMeshData->meshHasSeparatePartitionIDs);
+
+  semanticMeshData->collisionMeshData_.primitive = Mn::MeshPrimitive::Triangles;
+  semanticMeshData->updateCollisionMeshData();
+
+  if (semanticScene && (semanticScene->buildBBoxFromVertColors())) {
+    // FOR VERT-BASED OBB CALC build semantic (actually AABBs currently)
+    semanticMeshData->buildSemanticOBBs(semanticMeshData->cpu_vbo_,
+                                        semanticMeshData->objectIds_,
+                                        semanticScene->objects(), dbgMsgPrefix);
   }
 
+  ////
+  return semanticMeshData;
+}
+
+std::vector<std::unique_ptr<GenericSemanticMeshData>>
+GenericSemanticMeshData::partitionSemanticMeshData(
+    std::unique_ptr<GenericSemanticMeshData>& semanticMeshData) {
```

Comment:
Why does this take the `semanticMeshData` as a mutable argument? I don't see the function modifying it inside, so it could be `const&` instead.

Moreover -- similarly to my usual complaints about `Optional` overuse -- why the pointer wrapper? Just `const GenericSemanticMeshData&` would be enough as well, that way you also don't need to handle the case when a null pointer would get passed. Similarly, I think on the output it could also be just `std::vector<GenericSemanticMeshData>` -- unless there's something really specific done in the class, it should be relatively cheap to move.

Reply:
These functions return a smart pointer to the object because these are getting moved into the meshes_ map, which holds smart pointers of objects.